archive/logic_puzzles/knights_and_knaves.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2020 Dan Stanescu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dan Stanescu.
+-/
+
+import tactic
+
+/-!
+# Knights and knaves puzzles
+Two puzzles from "Knights and Knaves" by Raymond Smullyan.
+For an online description of these puzzles, see: http://mesosyn.com/mental1-6.html
+More information on R. Smullyan: https://en.wikipedia.org/wiki/Raymond_Smullyan
+
+There is an island where all inhabitants are either knights or knaves. 
+Knights always tell the truth. Knaves always lie.
+Joe and Bob are two inhabitants of this island.
+-/
+
+inductive islander
+| knight
+| knave
+
+notation `y` := islander.knight
+notation `n` := islander.knave
+
+structure Q := (Joe Bob : islander)
+
+/--
+In this first puzzle, only Joe makes a statement.
+He affirms that both protagonist (i.e. both Joe and Bob) are knaves.
+Question: what kind of islanders are they in fact?
+-/
+
+def Q1 := Q
+
+namespace Q1
+
+variables (q : Q1)
```

Comment:
Why introduce `q : Q1` at all? You could just have `variables (Joe Bob : islander)` in each namespace, I think.

Reply:
This is somehow the same question as before. I'd say it's just a matter of choice.

Reply:
Okay, but I think this may be part of what @robertylewis meant when he said "I'd love to see this in idiomatic Lean,". Perhaps we can see what he thinks?

Reply:
When I introduced `structure Q` and `(q : Q1)` in my son's file, my goal was to use `q.S1` instead of `S1 Joe Bob`, and I wanted this in order to save him some keystrokes (he can't type fast yet).

In case of knights and knaves I think that we should have something like `islander.says (i : islander) (p : Prop) := cond (i = knight) p (not p)`, then use `Joe.says` and `Bob.says`.

I wonder whether we can use some `fintype` magic to ask Lean to list all possible answers.